source/Engine.cpp

```diff
@@ -870,9 +866,29 @@ void Engine::Step(bool isActive)
 	// Draw crosshairs on any minables in range of the flagship's scanners.
 	double scanRange = flagship ? 100. * sqrt(flagship->Attributes().Get("asteroid scan power")) : 0.;
 	if(flagship && scanRange && !flagship->IsHyperspacing())
+	{
+		// Decide before looping whether or not to catalog asteroids.  This
+		// results in cataloging in-range asteroids roughly 3 times a second.
+		bool shouldCatalogAsteroids = !Random::Int(20);
+		shouldCatalogAsteroids &= !isAsteroidCatalogComplete;
+		bool scanComplete = true;
 		for(const shared_ptr<Minable> &minable : asteroids.Minables())
 		{
 			Point offset = minable->Position() - center;
+
+			// Autocatalog asteroid: Record that the player knows this type of asteroid is available here.
+			if(shouldCatalogAsteroids)
+			{
+				if(!Random::Int(10) && (minable->Position() - flagship->Position()).Length() <= scanRange)
+				{
+					asteroidsScanned.insert(minable->Name());
+					for(const auto &it : minable->Payload())
+						player.Harvest(it.first);
+				}
+				if(!asteroidsScanned.count(minable->Name()))
+					scanComplete = false;
+			}
+
 			if(offset.Length() > scanRange && flagship->GetTargetAsteroid() != minable)
```

Comment:
What's this nonsense? ik u didnt write it but we're assessing scan range based on the center of the system and not the distance to the flagship, as it should be.
If we change this we only have to calculate the distance once (using DistanceSquared and the scanRange modified as I mentionned)

Reply:
The center *is* the flagship position lol

Reply:
Oh well in that case we should be reusing that here, I just assumed if that had been the case it would have been used again here

Reply:
I wasn't sure so didn't want to rely on it